WIP: GCC 14.2 and macOS aarch64 builds

mcspr opened the pull request:
Only tested x86_64 builds & some gdb running. esp8266/Arduino requires some additional patching done to avoid newly introduced warnings & swserial has to be fixed yet again. I would assume everything works correctly, unless xtensa is actually broken (somehow) in the GCC right now

---

Like pico builder, CC=..., CXX=..., STRIP=... are replaced conditionally for mac aarch64 builds

Makefile changes also fix shadowing in LDFLAGS that was causing GDB config & make to misunderstand what libs go where. Also, ported ncurses & gmp & mpfr & expat builder approach to link up-to-date libs into the GDB binary statically.
(note libtinfo.a <-> libtermcap.a linkage; I believe it should be done in pico builder too instead of sed s/.../.../ing the makefile)

Most places use 'last-argument-wins' to substitute things for ./configure & make, or place a makefile task variable to replace the existing one

Flash strings matching ported from the old unreleased patch. Basically, compiler stores some known 'flash-string-type' prefixes for sections, so only `__attribute__((section(".irom0.pstr")))` (or, whatever is specified through compiler arg) is necessary from the program side. GCC some time in 13.x or 14.x made templates accept section attribute, so current approach may break in unexpected places when there is also assembly in the same string.
I may be missing some obvious linker script syntax or linker flags / options to change output section flags, though. 

Flash string in exceptions patch is replaced w/ a python-wrapped [ast-grep-py](https://pypi.org/project/ast-grep-py/) script that goes through every libstdc++ header and replaces standalone "..." in throwing functions into a separate symbol that can then have section attribute attached.

All of the previous \[xtensa\]-... patches dropped, as they are upstream. There are 13 more in the master, but can be added after this is done (if they don't break anything for the 14.2 as base)

'-mforce-l32' <s>is gone for right now, and the patches around it</s>
[ported version from 12.1](https://github.com/jcmvbkbc/gcc-xtensa/tree/xtensa-force-l32-12.1) w/ experimental `__attribute__((target("force-l32")))` & `pragma target` support; plus, fixed mforce-l32-strlen compiling into stdlib strlen

EH frame obj count replaced with the libstdc++ configure param that sets up static pool
*edit:* possible to replace with esp-idf patch that introduces weak symbol as configuration at build-time of the app
https://github.com/espressif/gcc/commit/f5e8a6560a932506a41236eabdd365f40fff9a91

---

Builds, some runtime tests

- [x] Linux x86\_64 builds toolchain & tools
- [x] MACOSARM builds toolchain & tools
- [x] MACOSX86
- [x] Windows 32 / 64
- [x] rpi
- [x] Linux 32 / aarch64

mcspr committed:
wip

mcspr committed:
sync with pico tools builder for macos, bump gdb n deps versions

mcspr committed:
libstdc++-v3 flash string patcher

using ast-grep-py to modify source headers directly, instead of
migrating previously used .patch as-is

add utility helper header into the distribution to also skip generic
macro define-undefine happening previously

mcspr committed:
Fix py3.8 / ubuntu 20.04 compat and call flash string patcher in the makefile

But, better option might be to update system python / install 3.9+,
since ast-grep-py requires rust toolchain present when pip-installing

mcspr committed:
Using up-to-date URLS from respective project home pages

Also support .xz archives when necessary

mcspr committed:
GCC static exception pool, remove custom patch

mcspr committed:
Allow to patch newlib for specific GCC version

mcspr committed:
GMP 6.2.x+ needs configure to generate headers

mcspr committed:
gcc5+ w/o cloog, use local libelf blob

mcspr committed:
Patching base

mcspr committed:
PSTR macros w/o asm magic, flash strings section flags helper

.ld does not appear to have any way to inject flags, only filter existing ones (from inputs)
asm does not appear to have any way to specify object size preemptively (...maybe it does?)

because section name is currently injected in template code, avoid clobbering its contents with inline assembly
inject flags programatically, so varasm.cc makes use of them and creates correct output asm at a later point

also, clean-up pgmspace.h from newlib. pending change directly in the newlib repo

mcspr committed:
Support std::__throw... (basic_string.h, etc.)

relax 'count' warning, but only for the string header
fix nesting issue with ast parser

mcspr committed:
invalid gcc config opt

mcspr committed:
Relocate GMP & MPFR to builddir

mcspr committed:
Replace only the first header anchor

basic string and bitset break otherwise

mcspr committed:
MACOS{ARM,X86} cannot build GMP assembly files

mcspr committed:
Shared deps, --prefix=... fix & configuration flags per env

partially synced ./configure & gdb deps changes from the pico builder
split some more deps and correctly fetch & unpack archived urls
(note that .lz aka lzip is not currently installed in the builder image)

ensure CC=... & CXX=... are explicitly set, triplet detection does not
work for some reason and the build falls back to /usr/bin/gcc
(which obviously does not work)

mcspr committed:
binutils MACOSARM builds

mcspr committed:
fixup patching order for older gcc

mcspr committed:
ncurses typo

mcspr committed:
Fix LDFLAGS=... shadowing & gdb static linkage

ref. pico builder, -ltermcap can be replaced with tinfo
drop unused code related to LDFLAGS and try to keep the rest to the env export

mcspr committed:
hal & tools MACOSARM builds

forcing CC=... CXX=... to reference actual host compiler, not gcc
forcing strip to be no-op for every possible invocation

earlephilhower commented:
> '-mforce-l32' is gone for right now, and the patches around it

IIRC, this was related to constant literals being only accessed by 32-bit reads so they would be safe for flash storage.  It's been a while, but looking at the patch it seems to do a l32 and then a shift to extract the appropriate bits in the machine definition.  If the core now will always require the unaligned read exception handler this may not be required.

mcspr commented:
Tnx, it does look like that, but I mostly do have trouble figuring out the order of the pattern -> constrains -> doing something to the code. Machine description in both patterns are quite dense. Have to rtfm more, [and documentation is pretty descriptive there](https://gcc.gnu.org/onlinedocs/gccint/Machine-Desc.html) :/

Especially, what exactly brings about
> "Memory that is not in a literal pool."

Two patches 'around' the mforce are dealing with string literals, apparently, since both reorganize strcmp in some way. I would guess type-info one is likely not very important when rrti is disabled. The other one is dealing with unwind, so it is something happening while exceptions are enabled.

*edit:* I do wonder if '-mstrict-align' is applicable here
(<s>TARGET_FORCE_L32 is just another code generation magic</s>)

jjsuwa-sys3175 commented:
> All of the previous [xtensa]-... patches dropped, as they are upstream. There are 13 more in the master, but can be added after this is done (if they don't break anything for the 14.2 as base)

Most of them are not related to ESP8266, but the last two

- [Define TARGET_DIFFERENT_ADDR_DISPLACEMENT_P target hook](https://github.com/gcc-mirror/gcc/commit/211459e12147e9ed42c0a0947b1b95b551e03ef3)
- [Fix the issue in "*extzvsi-1bit_addsubx"](https://github.com/gcc-mirror/gcc/commit/b433140a6cb40acedb2e6cb43c4e5a388e33f805)

should be applied. The latter is an obvious bug fix (for an actual problem with ESP32), so it must be.

mcspr committed:
Statically link mingw built tools with libgcc & libstdc++

mcspr committed:
Upstream xtensa patches

mcspr committed:
mingw inttypes.h missing SCNx8 support

> MS runtime  scanf appears to treat "hh" as "h"

and mingw does not ship a replacement

mcspr committed:
unused

earlephilhower commented:
I'm not sure if you're using the Docker setup I did for building things, but FYI for the latest GCC and binutils and MacOS/ARM stuff there's `earlephilhower/gcc-cross-v12`.  (Don't ask where v2 through v11 went...I don't remember!  :laughing: )

mcspr committed:
naming

mcspr committed:
ubuntu-20.04 image does not alias gcc to either ...-cc or ...-c++

mcspr committed:
dict.__len__ exists

mcspr commented:
> Most of them are not related to ESP8266, but the last two should be applied

attached

> I'm not sure if you're using the Docker setup I did for building things, but FYI for the latest GCC and binutils and MacOS/ARM stuff there's earlephilhower/gcc-cross-v12. (Don't ask where v2 through v11 went...I don't remember! 😆 )

I do, after noticing it in pico README file. Python dependency does not like the not-so-up-to-date python distributed with the 20.04 though, since pypi only provides binaries for python 3.9+. Either python needs a bump, or rust toolchain must be installed instead

earlephilhower commented:
Did not notice anything on my last run of PQT, but it's been ages since I looked at the EQT here.  Sorry!

If you're just talking about building the portable Python3 for Mac (for Win there's a download) then I would just punt on that and ship the older one for x86 Mac and use the [one](https://github.com/earlephilhower/pico-quick-toolchain/releases/download/4.0.1/python3.18-portable-darwin_arm64.tar.gz) I [build for ARM Macs](https://github.com/earlephilhower/pico-quick-toolchain/blob/master/.github/workflows/build-macarm-python.yml) in PQT since there's nothing toolchain specific.

mcspr commented:
> Did not notice anything on my last run of PQT, but it's been ages since I looked at the EQT here. Sorry!

Ah, sorry if I was not clear enough. Python used in the docker image is the issue for the Makefile and the libstd_flash_string_decls.py patching the libstdc++ includes at .stage.patch
I could pre-generate the .patch for the time being, though

> If you're just talking about building the portable Python3 for Mac (for Win there's a download) then I would just punt on that and ship the older one for x86 Mac and use the [one](https://github.com/earlephilhower/pico-quick-toolchain/releases/download/4.0.1/python3.18-portable-darwin_arm64.tar.gz) I [build for ARM Macs](https://github.com/earlephilhower/pico-quick-toolchain/blob/master/.github/workflows/build-macarm-python.yml) in PQT since there's nothing toolchain specific.

Makes sense, yep. I do wonder though... Why not ship statically linked python per-arch, even build it right in the builder?
At least some sources claim it is technically possible to statically a lot of internally used libs to reduce variability between builds - https://github.com/astral-sh/python-build-standalone - also meaning, python version stays the same for every platform

mcspr committed:
-mforce-l32

mcspr commented:
> Two patches 'around' the mforce are dealing with string literals, apparently, since both reorganize strcmp in some way. I would guess type-info one is likely not very important when rrti is disabled. The other one is dealing with unwind, so it is something happening while exceptions are enabled.

Also, to answer my own question... (hopefully?)

From the context of type_info `__builtin_strcmp` and unwinder `__builtin_strlen`, both are using boot rom 'strcmp' and 'strlen' respectively. In our environment, both char string arguments are usually w/ an address lacking unaligned read support. rom funcs apparently don't do anything about it, raising the unaligned read exception

> IIRC, this was related to constant literals being only accessed by 32-bit reads so they would be safe for flash storage

*After* applying -mforce-l32 makefiles patch to libsupc++ typeinfo & tinfo & libgcc unwinder, charptr arguments are dereferenced using those patched instructions. Hence the '_safe' suffix meaning 'safeness' b/c of the patch, not something safe about the code itself
(also, as a sideffect, gcc won't silently replace strlen & strcmp lookalikes in the code with the stdlib calls)

earlephilhower commented:
> Why not ship statically linked python per-arch, even build it right in the builder

Static or dynamic isn't the problem, it's the hardcoded `/usr/lib/python3/...` paths that are compiled into the binary.  In the portable version the `PREFIX`(equivalent) is `.` so you can just install it in any directory and it will find the support files relative to that, not using an absolute path.  The Python configuration script actually tries to disallow this config so that check needs to be [removed](https://github.com/earlephilhower/pico-quick-toolchain/blob/71f729a094edf4eca54a49ef40e4465f417d06d7/.github/workflows/build-macarm-python.yml#L28) to let the build work.

If the static versions are also portable and don't need any global paths we could ship them instead if they work with the serial control lib we use.  When the toolchain here was built I am pretty sure that repo didn't exist, and since we just need a basic Python install once it works there's no reason to update once a single Python portable package is available.

Python3 doesn't build for Macs using OSXCross.  The X86 version I actually built manually in a Hackintosh VM.  For the ARM one, GH provides ARM Mac runners so it's simple enough to build there (one time) using a GH Action and store the artifact.

Looks like I need to rebuild the Docker image with a new Ubuntu base (or an `apt-get update`) before Ubu20.04 goes EOL in 30 days.  And make sure I install any `pip` packages so even if the PyPi support ends, the image will continue working.

I experimented w/Ubu22.04 but its `glibc` is not backwards-compatible w/20.04 so the Linux toolchain would not be compatible w/20.04 systems.  :cry:

mcspr committed:
fix -mflash-string-section=... initialization

move back to the main xtensa.cc and really really make sure defaults get set up
reordered init of the prefixes slightly earlier to avoid -mforce-l32 patch conflict

mcspr committed:
port gcc-stdcsupp-safe.patch & gcc-unwind-safe.patch

typeinfo header __attribute__((__target___("force-l32"))) to compile outside of the lib
updated strlen implementation char -> unsigned char so the pattern picks it up too

(TODO is it necessary to check the upper func when inlining force-l32 targets?)

mcspr committed:
re-re-fix task: VAR1=... VAR2=...

makefile parses variables as CC=...rest-of-the-line...
set one variable per line for every tool task

mcspr committed:
libexpat w/o manpage and xmlwf tool

sometimes fails, and not necessary

mcspr committed:
-mforce-l32 fixed constraint & patch ordering

mcspr committed:
task stamps & order

reset and clean is delegated per dir, don't do it twice
checkout mkspiffs and mklittlefs correctly, only one has modules

mcspr committed:
signal when gcc make fails

mcspr committed:
lx106-hal builds w/ gcc-cross-v12 never make split obj files

also fix empty log.stage.patch appearing in the lx106-hal repo dir

mcspr committed:
arduino update & build fixups

arduino stopped shipping tools/sdk/libc, just c/p the lib obj
arduino-specific tasks working on the current tree
tarball hash & date known at $(MAKE) invocation, dont use globbing

mcspr committed:
missing mkdir repodir when downloading & prune unused / misnamed tasks

missing branch version pins

missing log overwrite at the start of the stage

repodir must exist for checkout & blobs & patch
blobs needs checkout first b/c of libelf
patch needs both checkout and blobs

attempt to fix preparation tasks stamps so the stage always runs
prefer that 'pre' tasks are always running, checking internally whether
something has to execute and doing nothing otherwise

mcspr committed:
clean-up libstdc++ patcher

explicitly check when attempting to patch helper header

correctly using root path in the libstdc++ patcher

include .tcc as a valid suffix

add _M_check_... variants to checked signatures

call expr regex checking for std:: & this-> prefixes

fix nesting check for __N(...) & actually modify replacement text

mcspr committed:
macos always disables lto builds

same restriction for x86_64 and aarch64

mcspr commented:
> I experimented w/Ubu22.04 but its glibc is not backwards-compatible w/20.04 so the Linux toolchain would not be compatible w/20.04 systems. 😢

fwiw, builds above surely work ok on the 20.04 system. the trick is to just avoid glibc, bundling musl instead

they do patch out that specific absolute path requirement, plus some additional stuff ([ref cpython-unix/*.patch in the astral-sh repo](https://github.com/astral-sh/python-build-standalone/tree/main/cpython-unix))
install/bin/python3 uses relative path to figure out the libraries location.
tarball with install/ still has a flat directory of stdlib site-packages, but it can be repackaged as python313.zip

-I aka isolated mode still a requirement, ~/.local/lib/python$VER/site-packages should be ignored even with these builds

---

On the topic, -mforce-l32 back. Additionally, `__attribute__((target("force-l32")))` is implemented to flag header code

While normal exception code seems to work, something like e.g. std::logic_error (tracing std::vector::at(...) throw) does some formatting and attempts to construct a std::string member. This obviously breaks when attempting to read flash address, but cannot be easily fixed w/ 'force-l32' sprinkled everywhere without replacing `__builtin_strlen` expression itself.

Other that just ignoring it for status quo, I would imagine the most straightforward fix is to alias functions in question (memcmp, memcpy, strlen, strncmp, strncpy and strstr) to their `_P` variants.
For user code that means less headache when dealing with various containers. Wonder why this was not implemented like this originally.

Quick check of SDK, there are mixed references to both `ets_memcpy` and `memcpy` (also `memcmp` and `strncmp`) PROVIDE'd via ROM.
*edit:* `etc_...` internally call these raw funcs, so there's no difference between them
Could use `rom_...` alias to still provide them, if there's a good reason to still have them outside of `ets_...` prefix

There is also an issue with `__gnu_cxx::__snprintf_lite`, but it can actually use the 'force-l32' helper (and not constexpr)
unlike 10.3, libstdc++ is not 'verbose', fmt throwing is not enabled by default (?)

---

Apparently, neither macOS flavours support --enable-lto 
Specific error is with libtool script & its attempt to call `nm -B`, which is not implemented for osxcross tool. LTO:=false is default, though, and not sure if it even works for other targets
*edit:* ...and ./configure reports 'no lto support' for osxcross 'ar', so there's that

I'll have to re-check CC=$(TRIPLE)-cc and CXX=$(TRIPLE)-c++ assignment, but seems like a reasonable default for tools and gcc1 builds

mcspr committed:
non-cplusplus-only newlib

mcspr committed:
install libhal.a to xtensa-lx106-elf/lib

same as libstdc++.a & libstdc++-nox.a, allowing -lhal to work
TODO automake could somehow redirect there via tooldir=... like gcc does?

mcspr commented:
> Other that just ignoring it for status quo, I would imagine the most straightforward fix is to alias functions in question (memcmp, memcpy, strlen, strncmp, strncpy and strstr) to their _P variants.
For user code that means less headache when dealing with various containers. Wonder why this was not implemented like this originally.

so, some side-changes...
https://github.com/esp8266/Arduino/compare/master...mcspr:gcc14/wip?expand=1
https://github.com/earlephilhower/newlib-xtensa/compare/xtensa-4_0_0-lock-arduino...mcspr:newlib-xtensa:gcc14/string-h?expand=1

newlib forced to adhere 'rom\_...' / 'newlib\_...' / 'xtensa string.h' separation
should be checked extensively, though. pretty worried about it short circuiting somewhere
`..._P` prefers to always do the `pgm_...` stuff, flash-addr check relegated to the libc level

user side just needs symbol renames in SDK & .ld for the actual code
dumb example w/ exceptions compiles just fine and constructs `std::string x(PSTR("hello there"));` without issues. std::logic\_error throws correctly as the result.

mcspr committed:
missing patches & correct newlib branch

mcspr committed:
experiment w/ gcc-func-sect.patch replacement

whenever gcc creates internal string array, make sure it knows where it
came from originally. only matters when constants-merge flag is active

mcspr committed:
libcurses.a -> libncurses.a

while --disable-widec is used, gdb should not test system lib at all

mcspr commented:
- https://github.com/earlephilhower/esp-quick-toolchain/pull/54/commits/f4edca2d13e2a488f3c2abe3a8b9ce3085224417 - note that newlib refers to mcspr/newlib-xtensa branch. strcasestr & memmem are also reorganized to properly work when *both* or either one of haystack and needle are in flash (...like memchr already works, so might as well)

- https://github.com/earlephilhower/esp-quick-toolchain/pull/54/commits/26ce4f7b0913b822a41068989847db3fa1e2a081 - instead of reverting the commit mentioned in #9, internal string arrays are manually handled to include originating 'identifier'. in case of `__func__`-like stuff and additionally `__builtin_{FILE,FUNCTION}();`
I'm fairly certain it works :) At least scanning for `TREE_CHAIN (TREE_TYPE (` / `TREE_CHAIN (type)` does not seem to reveal any possible collisions. It is used for function types, but not arrays as far as I can tell (char[] in this case). Internal tests related to `__func__` do still pass, but I have to figure out filter out unrelated platform-specific tests which don't work b/c of `int main()` requirement. Will have to use host build, probably...

  It also builds, at least allowing builtins to go to the flash for my use-case. But, have to re-check where there are a lot of asserts (or something similar)

mcspr committed:
section dot placement fix for missing -fmerge-constants -ffunction-sections

remove extra comments, commit msg would suffice

mcspr committed:
autocorrect lx106-hal install per target via exec-prefix; also reuse target arch directory var for others

per https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Installation-Names.html

as it never seem to pick up on the cross-build target thing
(which probably requires to edit configure.ac & autoconf after)

mcspr committed:
$TARGET_ARCH, formatting

mcspr committed:
formatting

remove install newlib cfg from 48f7b08d1679365ea462675dd3bfb9d83c9bf11a
remove install libgcc commented out

mcspr committed:
missing variant what() static string

mcspr committed:
don't forget to replace string literals in .cc sources

fix search anchors in edited data, not the original
multiple search paths as arg, explicit helper header path arg

mcspr commented on Makefile:
```diff
@@ -566,52 +803,86 @@ clean: .cleaninst.LINUX.clean .cleaninst.LINUX32.clean .cleaninst.WIN32.clean .c
+
+.stage.LINUX.arduino-toolchain:
+	echo "-------- Copying GCC and LIBSTDC++ libs"
+	cp -vu $(call install,$@)/$(TARGET_ARCH)/lib/libstdc++-exc.a $(ARDUINO)/tools/sdk/lib/.
+	cp -vu $(call install,$@)/$(TARGET_ARCH)/lib/libstdc++.a     $(ARDUINO)/tools/sdk/lib/.
 	echo "-------- Copying toolchain directory"
-	rm -rf $(ARDUINO)/tools/sdk/xtensa-lx106-elf
-	cp -a $(call install,$@)/xtensa-lx106-elf $(ARDUINO)/tools/sdk/xtensa-lx106-elf
+	rm -rf $(ARDUINO)/tools/sdk/$(TARGET_ARCH)
+	cp -va $(call install,$@)/$(TARGET_ARCH) $(ARDUINO)/tools/sdk/$(TARGET_ARCH)
+
+.stage.LINUX.arduino-hal:
+	echo "-------- Copying HAL lib"
+	cp -vu $(call install,$@)/$(TARGET_ARCH)/lib/libhal.a $(ARDUINO)/tools/sdk/lib/.
```
any reason for .a copies distributed with release or git (aka tools/get.py)?
the ones in the tools/sdk/lib actually take priority based on search path injected in both PIO and platform.txt

afaik, all three are linking just fine from {toolchain}/xtensa-lx106-elf/lib already
both pio & ide

earlephilhower replied:
These were part of the core from before I started working on it.  But as I remember it, there was a lot of churn in the libraries and it was thought that just updating the core would be better for folks than requiring whole new downloads of the toolchain every release.  It was also much faster since you didn't need to rebuild all 9 architecture GCCs, just the native Linux one and copy in the libs.

mcspr commented on Makefile:
```diff
@@ -126,23 +126,79 @@ else ifeq ($(GCC), 11.1)
+    GCC_PKGREL    := 140200
+    GCC_REPO      := https://gcc.gnu.org/git/gcc.git
+    GCC_DIR       := gcc-gnu
+    BINUTILS_BRANCH := binutils-2_44
```
there is an issue going back to binutils 2.37 where statically built distribution contains /lib/bfd-plugins/libdep.a instead of /lib/bfd-plugins/libdep.so
https://sourceware.org/bugzilla/show_bug.cgi?id=27082

initially, this was noticed in mingw. but, they just opt out of any plugins. plus, upstream does not take account of libdep.a, only libdep.dll.a
https://github.com/msys2/MINGW-packages/blob/68f7d4665c396a464536871b1de7b680a47a8fa7/mingw-w64-binutils/PKGBUILD#L150-L151

/lib/bfd-plugins/libdep.a happens to be in the output for every variant here, though, not only mingw one.
libtool invocations in binutils do tag it as '--tag=disable-static', but it does not prevent .a creation (which I would assume it would)
https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=ld/Makefile.in;h=3e24bd4deb26f787439c69e57f9321755f2d258a;hb=815d9a14cbbb3b81843f7566222c87fb22e7255d#l1066

(nb. when lto is enabled, gcc also builds and links /lib/bfd-plugins/liblto_plugin.so)

earlephilhower replied:
The way the `weak` attribute is used here, I don't think that LTO is really an option.  So you may want to turn that off and save the effort.  On the Pico we don't use it and there are no issues across MinGW or other archs.

mcspr replied:
I meant the installation method, LTO is disabled so only plugin is libdep w/ the invalid binary contents
libdep meant to embed '-L...' metadata into archives, which is then sometimes helpful to the linker

Note that Pico also ships libdep archive, but it is not loaded for whatever reason when I build empty sketch
`C:\Users\maxim\AppData\Local\Arduino15\packages\rp2040\tools\pqt-gcc\4.0.1-8ec9d6f\lib\bfd-plugins\libdep.a`
`ar.exe` & `ld.exe` still mention plugins, so support was built in (`--enable-plugins` implicitly enabled)
Again, comparing to lto-plugin which does warn about static & shared, binutils repo probably should not have embedded them into ld/ directory and had a separate one which would yell about incompatible opts

https://github.com/earlephilhower/esp-quick-toolchain/pull/54/commits/2b4c583eb79ceb45de1ff3accb67b10005325330 just rebuilds ld directory with dynamic options and c/p the plugin .so / .dll

mcspr committed:
gcc link after gcc install

mcspr committed:
attempt to fix binutils dynamic plugins loader

rebuild binutils-gdb/ld w/ --enable-shared after building normally
apply fix from mingw and remove /lib/bfd-plugins/libdep.a

mcspr committed:
dumpvars per stage

mcspr committed:
fix typos & dont depend on directory without a rule

don't depend on sub-directory without any rule creating it
shorter tar pattern, it already recognizes filetypes
shell ${{ -> $${ (this is not github actions)

mcspr committed:
mklittlefs-4.0.2 w/ littlefs-2.5.1

mcspr committed:
only gcc14.2 w/ binutils>2.37

mcspr committed:
hal still needs CC=... when building via osxcross

explicit override seems to mess up the target-build-host

mcspr committed:
libdep plugin is only gcc14 right now

mcspr commented:
mklittlefs local branch reverts to 2.5.1 for the time being
https://github.com/mcspr/mklittlefs/tree/4.0.2-littlefs251
the amount of other stuff broken warrants v4.0.0 for esp8266 though anyway. might as well update it too

rebased arduino-cli tests & a lot of backlogged patches seem to work
https://github.com/mcspr/esp8266-Arduino/actions/runs/15009572288
(only arm macos toolchain added to package .json, plus the python3.18 from the pico)

would have to build toolchains manually for the rest, since the Actions CI hangs somewhere in .stage.blobs :/
.stage.gitclone also takes ~10min, and pretty sure it would not fit into cache
(maybe on non-free tier it works though)
*edit:*  ...and actually try the package .json sync tool this time

earlephilhower commented:
For mklittlefs, if you update it then you need to update the core.  There was a LittleFS on-flash format change and it's not compatible with the older format.  The new code will convert old FSes to the new format on first mount.

Re: building, it needs ~60GB of disk space and runs for ~40 mins on my 6T12C Ryzen 5600X @ 32G with `-j16` (higher than thread count because there are lots of non-CPU intensive scripting stages as part of building).  It might be possible to  use chained GH actions, but I doubt it.  And with a 2C limit per VM it would be pretty painful.

mcspr commented:
Huh, thanks for the tip about overcommiting to nproc. It still takes almost 10min per flavour with a lazily configured laptop setup, but it'll have to do. And I also haven't noticed the 48gig build dir, that would indeed be even more problematic for githubs ci without some clever tweaks

Regarding littlefs, that was the reason for the temporary downgrade.
esp8266 repo has some on-device tests, but those depend on spiffs. Will attempt to port & run those based on littlefs in case I missed something in the newlib. 

Having also thought about bumping newlib to 4.5.0 snapshot, but would have to dig through the build system changes as well as newlib/sys/xtensa -> libgloss move made by espressif some time before 4.4.0

mcspr commented:
> Having also thought about bumping newlib to 4.5.0 snapshot, but would have to dig through the build system changes as well as newlib/sys/xtensa -> libgloss move made by espressif some time before 4.4.0

Not much changes code-wise, as it turns out. Also removes some unused changes previously done to the tree in the old newlib-xtensa patch. Will switch out after some more checking
https://github.com/mcspr/newlib-xtensa/commit/5238b87265cd3791fa5f9f01614446a1c5309ab2

libgloss/ is now shipping no-op crt0.o & some no-op syscalls. Vendor header is expecting ESP32, so that had to be more like the version in lx106-hal. Probably makes sense to keep it in the tree vs. merging w/ the whole lx106-hal header

libc/sys was originally used but later moved in Espressif branch to libgloss/ & libc/machine/
I am pretty sure it's the only sensible place for real overrides + pgmspace. Only significant change is splitting one big string\_pgmspace.c into separate files to avoid linking the whole .o, reducing the total binary size

libc/machine is more extensively used for libc.a replacements, as I've missed the fact that the build system *appends* to `libc_a_SOURCES`. Because of the ordering used in the top-level Makefile, same-name files in libc/machine/xtensa/ dir replace the ones libc/ owns. That removes the need for `#if 0` wraps in the newlib-xtensa impl, or the `newlib_...` renames I originally did.

Only strange issue is nano-printf suddenly not depending on wide-char `_fputwc_r` since
https://github.com/mcspr/newlib-xtensa/commit/3b97a5ec67a5a52c130158bb143949cd842de305

Actual condition was changed previously in https://github.com/mcspr/newlib-xtensa/commit/0239bc706a32364edc95ea06f667f4b233f74b77
so it can be easily disabled with `--disable-newlib-wide-orient` (afaik, wide chars are not expected to be streamed anywhere)
or, `--enable-newlib-elix-level=1`, but that also disables some other useful string funcs

earlephilhower commented:
In case you haven't seen it, some very recent mailing list stuff on the `wchar_t` undefined functions:  https://sourceware.org/pipermail/newlib/2025/021732.html  .

mcspr commented:
So the thread above also mentions this error, which I've not seen just yet
https://sourceware.org/pipermail/newlib/2025/021818.html
```
  libc_nano.a(libc_a-wcsftime.o): in function `__strftime':
  /builds/idf/crosstool-NG/.build/riscv32-esp-elf/src/newlib/newlib/libc/time/../time/strftime.c:879:(.text.__strftime.isra.0+0x2ac): undefined reference to `swprintf'
```

This actually references wcsftime.c being built with `#include "strftime.c"`, but not exactly sure what calls `wcsftime` and where
*edit* looking at libstdc++, at least one source of it is `std::locale` and anything attempting to use it :/

https://sourceware.org/git/?p=newlib-cygwin.git;a=blob;f=newlib/libc/time/strftime.c;h=56f227c5f3b8e4ebcb2f0dc25c4e26cf6e512bf6;hb=refs/heads/main#l317
https://sourceware.org/git?p=newlib-cygwin.git;a=blob;f=newlib/libc/time/wcsftime.c;h=5dde213004f9884571e327923f59a6041f971994;hb=refs/heads/main

Following that logic, newlib/libc/time/Makefile.inc *should* exclude wcstime.c based on `NEWLIB_NANO_FORMATTED_IO`

earlephilhower commented:
It kind of turned into a mess as I was digging into that, so in the end I simply moved off of nano-printf.  Drop the 
````
CONFIGURENEWLIBCOM += --enable-newlib-nano-formatted-io
````
and add in
````
CONFIGURENEWLIBCOM += --enable-newlib-io-c99-formats
CONFIGURENEWLIBCOM += --enable-newlib-io-long-long
````
Also need to remove the `-u printf_float -u scanf_float` from `platform.txt` and the Platform.IO build script.

On the ARM Thumb this change adds about 20KB to the final image.  I'm not really seeing any difference in runtime heap...both nano and normal do malloc/frees of small chunks as they work.

AIUI the nano printf was used on the ESP8266 from the beginning because the original boards came with only 256K or 512K of flash.  With the blob overhead, every bit of flash was important.  Now practically every 8266 board comes with 4MB flash devices, meaning that small savings isn't really a big deal.

mcspr commented:
Got around the same size increase for a no-float printf, 16KiB. `std::regex` and `std::locale` formatting increases by 23KiB.

So, the main difference here is full C99 compat - long long and `wchar_t`.
Is `wchar_t` support correct here, though? (`L'....'` literals, inner overloads for libc streams, etc.).   While it builds, `wchar_t` is u16 (like Windows) and not u32 (like Linux). Plus, a lot of neighbours expect `setlocale` and deal with different formatting of time / money / numbers / etc., which includes differently sized stdio streams. It is good for 'full' C99, but not sure how actually useful it is.

*edit:* notably, GCC / libstdc++ has `--disable-wchar_t`. newlib is still required to provide the type, its just none of the internals would try to use wide character support
https://gcc.gnu.org/onlinedocs/libstdc%2B%2B/manual/configure.html /--enable-wchar\_t/
*edit2:* ...but, GCC is the one providing `wchar_t` through `stddef.h`. hm.

earlephilhower commented:
>So, the main difference here is full C99 compat - long long and wchar_t.

I know I added half half and think I did long long to the nano-printf as part of the patches in the repo.

Anyway, for me the main difference is "it builds and runs existing code without me needing to rewrite its Makefile or time formatting code" :laughing: ).  

FWIW on the ARM wchar_t is 4 bytes...maybe there's an Xtensa machine definition thing going on to cause 16b.  AIUI this comes from GCC itself, not newlib.  I see in stddef.h:
````
#ifndef __WCHAR_TYPE__
#define __WCHAR_TYPE__ int
#endif
````

As an alternative, if you just want to get things out, would be to stick with Newlib 4.3.0 which we used for a year or so without incident.  Since then there have been some trigonometric function errors corrected and RISC-V updates which are probably not a big deal on the 8266.

mcspr commented:
True, but everything still has to be preferably going through the 32bit access macros, which means code & build gets edited either way 😭 
OK, I'll stop the rambling for now and try some actual tests during the week.

earlephilhower commented:
FWIW there have been a couple proposed patches on the Newlib mailing list just this last week related to the nano-printf and widechar issue.  Might want to just sit and see how it falls out in the next week or two.  Coupld possibly just bump, then, to the working HEAD...

mcspr commented:
> FWIW there have been a couple proposed patches on the Newlib mailing list just this last week related to the nano-printf and widechar issue. Might want to just sit and see how it falls out in the next week or two. Coupld possibly just bump, then, to the working HEAD...

btw rebasing newlib-xtensa base is annoying b/c of diverging commits. ['newlib-2_2_0'] commit (https://github.com/mcspr/newlib-xtensa/commit/7e72fffc61b29325ac5a875c5318cdf2f294ae28) is missing at upstream https://sourceware.org/git/newlib-cygwin.git main branch, but present in the tree.
(https://sourceware.org/legacy-ml/newlib/2015/msg00313.html ?)

I've broken 4.5.0 for the 4th time already, so it is yet again merging through various changes. Could also lose the history, and just base off of the upstream...
